src/server/game/Achievements/AchievementMgr.cpp

```diff
@@ -630,11 +670,385 @@ void PlayerAchievementMgr::SendPacket(WorldPacket const* data) const
+        if (CharTitlesEntry const* titleEntry = sCharTitlesStore.LookupEntry(titleId))
+            _owner->GetPlayer()->SetTitle(titleEntry);
+
+    // mail
+    if (reward->SenderCreatureId)
+    {
+        MailDraft draft(uint16(reward->MailTemplateId));
+
+        if (!reward->MailTemplateId)
+        {
+            // subject and text
+            std::string subject = reward->Subject;
+            std::string text = reward->Body;
+
+            LocaleConstant localeConstant = _owner->GetPlayer()->GetSession()->GetSessionDbLocaleIndex();
+            if (localeConstant != LOCALE_enUS)
+            {
+                if (AchievementRewardLocale const* loc = sAchievementMgr->GetAchievementRewardLocale(achievement))
+                {
+                    ObjectMgr::GetLocaleString(loc->Subject, localeConstant, subject);
+                    ObjectMgr::GetLocaleString(loc->Body, localeConstant, text);
+                }
+            }
+
+            draft = MailDraft(subject, text);
+        }
+
+        SQLTransaction trans = CharacterDatabase.BeginTransaction();
+
+        Item* item = reward->ItemId ? Item::CreateItem(reward->ItemId, 1, _owner->GetPlayer()) : NULL;
+        if (item)
+        {
+            // save new item before send
+            item->SaveToDB(trans);                               // save for prevent lost at next mail load, if send fail then item will deleted
+
+                                                                 // item
+            draft.AddItem(item);
+        }
+
+        draft.SendMailTo(trans, _owner->GetPlayer(), MailSender(MAIL_CREATURE, uint64(reward->SenderCreatureId)));
+        CharacterDatabase.CommitTransaction(trans);
+    }
+}
+
+bool AccountAchievementMgr::ModifierTreeSatisfied(uint32 modifierTreeId) const
+{
+    return AdditionalRequirementsSatisfied(sCriteriaMgr->GetModifierTree(modifierTreeId), 0, 0, nullptr, _owner->GetPlayer());
+}
+
+void AccountAchievementMgr::SendCriteriaUpdate(Criteria const* criteria, CriteriaProgress const* progress, uint32 timeElapsed, bool timedCompleted) const
+{
+    WorldPackets::Achievement::AccountCriteriaUpdate accountCriteriaUpdate;
+
+    accountCriteriaUpdate.Progress.Id = criteria->ID;
+    accountCriteriaUpdate.Progress.Quantity = progress->Counter;
+    accountCriteriaUpdate.Progress.Player = _owner->GetBattlenetAccountGUID();
+
+    accountCriteriaUpdate.Progress.Date = progress->Date;
+    accountCriteriaUpdate.Progress.TimeFromStart = timeElapsed;
+    accountCriteriaUpdate.Progress.TimeFromCreate = 0;
+    accountCriteriaUpdate.Progress.Flags = 0;
+    if (criteria->Entry->StartTimer)
+        accountCriteriaUpdate.Progress.Flags = timedCompleted ? 1 : 0; // 1 is for keeping the counter at 0 in client
+
+    SendPacket(accountCriteriaUpdate.Write());
+}
+
+void AccountAchievementMgr::SendCriteriaProgressRemoved(uint32 criteriaId)
```

Comment:
Unless you plan to add something within the brackets below,
commenting out `criteriaId` is recommended (travis-ci).